File: api/v1/schemas/admin_provider_circuits.py
# -*- coding: utf-8 -*-
"""Schemas for read-only admin provider circuit diagnostics."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class _AdminProviderCircuitModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class ProviderCircuitStateItem(_AdminProviderCircuitModel):
    provider: str
    provider_category: Optional[str] = Field(default=None, alias="providerCategory")
    route_family: Optional[str] = Field(default=None, alias="routeFamily")
    state: str
    reason_bucket: Optional[str] = Field(default=None, alias="reasonBucket")
    cooldown_until: Optional[str] = Field(default=None, alias="cooldownUntil")
    operator_action_ref: Optional[str] = Field(default=None, alias="operatorActionRef")
    created_at: Optional[str] = Field(default=None, alias="createdAt")
    updated_at: Optional[str] = Field(default=None, alias="updatedAt")


class ProviderCircuitEventItem(_AdminProviderCircuitModel):
    provider: str
    provider_category: Optional[str] = Field(default=None, alias="providerCategory")
    route_family: Optional[str] = Field(default=None, alias="routeFamily")
    event_type: str = Field(alias="eventType")
    from_state: Optional[str] = Field(default=None, alias="fromState")
    to_state: Optional[str] = Field(default=None, alias="toState")
    reason_bucket: Optional[str] = Field(default=None, alias="reasonBucket")
    request_count_bucket: Optional[str] = Field(default=None, alias="requestCountBucket")
    duration_bucket_ms: Optional[int] = Field(default=None, alias="durationBucketMs")
    failure_count_bucket: Optional[str] = Field(default=None, alias="failureCountBucket")
    operator_action_ref: Optional[str] = Field(default=None, alias="operatorActionRef")
    created_at: Optional[str] = Field(default=None, alias="createdAt")


class ProviderQuotaWindowItem(_AdminProviderCircuitModel):
    provider: str
    provider_category: Optional[str] = Field(default=None, alias="providerCategory")
    route_family: Optional[str] = Field(default=None, alias="routeFamily")
    window_type: str = Field(alias="windowType")
    window_start: str = Field(alias="windowStart")
    window_end: str = Field(alias="windowEnd")
    request_count: int = Field(default=0, alias="requestCount")
    reserved_units: int = Field(default=0, alias="reservedUnits")
    consumed_units: int = Field(default=0, alias="consumedUnits")
    released_units: int = Field(default=0, alias="releasedUnits")
    rejected_count: int = Field(default=0, alias="rejectedCount")
    success_count: int = Field(default=0, alias="successCount")
    failure_count: int = Field(default=0, alias="failureCount")
    timeout_count: int = Field(default=0, alias="timeoutCount")
    provider_429_count: int = Field(default=0, alias="provider429Count")
    provider_403_count: int = Field(default=0, alias="provider403Count")
    fallback_count: int = Field(default=0, alias="fallbackCount")
    probe_count: int = Field(default=0, alias="probeCount")
    cache_only_count: int = Field(default=0, alias="cacheOnlyCount")
    stale_served_count: int = Field(default=0, alias="staleServedCount")
    created_at: Optional[str] = Field(default=None, alias="createdAt")
    updated_at: Optional[str] = Field(default=None, alias="updatedAt")


class ProviderProbeEventItem(_AdminProviderCircuitModel):
    provider: str
    provider_category: Optional[str] = Field(default=None, alias="providerCategory")
    route_family: Optional[str] = Field(default=None, alias="routeFamily")
    probe_type: str = Field(alias="probeType")
    probe_source: str = Field(alias="probeSource")
    result_bucket: str = Field(alias="resultBucket")
    duration_bucket_ms: Optional[int] = Field(default=None, alias="durationBucketMs")
    created_at: Optional[str] = Field(default=None, alias="createdAt")


class ProviderRecentErrorBucketItem(_AdminProviderCircuitModel):
    reason_bucket: str = Field(alias="reasonBucket")
    count_bucket: str = Field(alias="countBucket")
    latest_at: Optional[str] = Field(default=None, alias="latestAt")


class ProviderSlaReadinessItem(_AdminProviderCircuitModel):
    provider: str
    provider_category: Optional[str] = Field(default=None, alias="providerCategory")
    route_family: Optional[str] = Field(default=None, alias="routeFamily")
    observed_since: str = Field(alias="observedSince")
    readiness_state: str = Field(alias="readinessState")
    reason_code: str = Field(alias="reasonCode")
    credential_state: str = Field(alias="credentialState")
    live_providers_enabled: bool = Field(default=False, alias="liveProvidersEnabled")
    provider_enabled: bool = Field(default=False, alias="providerEnabled")
    credentials_present: bool = Field(default=False, alias="credentialsPresent")
    dry_run_enabled: bool = Field(default=False, alias="dryRunEnabled")
    live_http_calls_enabled: bool = Field(default=False, alias="liveHttpCallsEnabled")
    broker_order_path_enabled: bool = Field(default=False, alias="brokerOrderPathEnabled")
    portfolio_mutation_path_enabled: bool = Field(default=False, alias="portfolioMutationPathEnabled")
    tradeable_data: bool = Field(default=False, alias="tradeableData")
    latency_bucket_ms: Optional[int] = Field(default=None, alias="latencyBucketMs")
    latency_state: str = Field(default="unknown", alias="latencyState")
    error_rate: Optional[float] = Field(default=None, alias="errorRate")
    error_state: str = Field(default="unknown", alias="errorState")
    freshness_seconds: Optional[int] = Field(default=None, alias="freshnessSeconds")
    freshness_state: str = Field(default="unknown", alias="freshnessState")
    recent_errors: List[ProviderRecentErrorBucketItem] = Field(default_factory=list, alias="recentErrors")
    circuit_advisory_state: str = Field(default="healthy", alias="circuitAdvisoryState")
    circuit_state_candidate: str = Field(default="closed", alias="circuitStateCandidate")
    live_enforcement: bool = Field(default=False, alias="liveEnforcement")
    would_block_call: bool = Field(default=False, alias="wouldBlockCall")
    would_change_provider_order: bool = Field(default=False, alias="wouldChangeProviderOrder")
    would_change_fallback_behavior: bool = Field(default=False, alias="wouldChangeFallbackBehavior")
    no_external_calls: bool = Field(default=True, alias="noExternalCalls")
    provider_behavior_changed: bool = Field(default=False, alias="providerBehaviorChanged")
    market_cache_behavior_changed: bool = Field(default=False, alias="marketCacheBehaviorChanged")


class ProviderCircuitDiagnosticsMetadata(_AdminProviderCircuitModel):
    read_only: bool = Field(default=True, alias="readOnly")
    no_external_calls: bool = Field(default=True, alias="noExternalCalls")
    live_enforcement: bool = Field(default=False, alias="liveEnforcement")
    provider_behavior_changed: bool = Field(default=False, alias="providerBehaviorChanged")
    market_cache_behavior_changed: bool = Field(default=False, alias="marketCacheBehaviorChanged")
    data_sources: List[str] = Field(default_factory=list, alias="dataSources")
    limit: int = 100
    redaction: List[str] = Field(default_factory=list)
    filters: Dict[str, Any] = Field(default_factory=dict)


class ProviderCircuitStatesResponse(_AdminProviderCircuitModel):
    generated_at: str = Field(alias="generatedAt")
    items: List[ProviderCircuitStateItem] = Field(default_factory=list)
    metadata: ProviderCircuitDiagnosticsMetadata


class ProviderCircuitEventsResponse(_AdminProviderCircuitModel):
    generated_at: str = Field(alias="generatedAt")
    items: List[ProviderCircuitEventItem] = Field(default_factory=list)
    metadata: ProviderCircuitDiagnosticsMetadata


class ProviderQuotaWindowsResponse(_AdminProviderCircuitModel):
    generated_at: str = Field(alias="generatedAt")
    items: List[ProviderQuotaWindowItem] = Field(default_factory=list)
    metadata: ProviderCircuitDiagnosticsMetadata


class ProviderProbeEventsResponse(_AdminProviderCircuitModel):
    generated_at: str = Field(alias="generatedAt")
    items: List[ProviderProbeEventItem] = Field(default_factory=list)
    metadata: ProviderCircuitDiagnosticsMetadata


class ProviderSlaReadinessResponse(_AdminProviderCircuitModel):
    generated_at: str = Field(alias="generatedAt")
    items: List[ProviderSlaReadinessItem] = Field(default_factory=list)
    metadata: ProviderCircuitDiagnosticsMetadata

File: apps/dsa-web/src/api/adminProviderCircuits.ts
import apiClient from './index';
import { toCamelCase } from './utils';

export interface ProviderCircuitDiagnosticsParams {
  provider?: string;
  routeFamily?: string;
  state?: string;
  eventType?: string;
  since?: string;
  limit?: number;
}

export interface ProviderCircuitDiagnosticsMetadata {
  readOnly: boolean;
  noExternalCalls: boolean;
  liveEnforcement: boolean;
  providerBehaviorChanged: boolean;
  marketCacheBehaviorChanged: boolean;
  dataSources: string[];
  limit: number;
  redaction: string[];
  filters: Record<string, unknown>;
}

export interface ProviderCircuitStateItem {
  provider: string;
  providerCategory?: string | null;
  routeFamily?: string | null;
  state: string;
  reasonBucket?: string | null;
  cooldownUntil?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface ProviderCircuitEventItem {
  provider: string;
  providerCategory?: string | null;
  routeFamily?: string | null;
  eventType: string;
  fromState?: string | null;
  toState?: string | null;
  reasonBucket?: string | null;
  requestCountBucket?: string | null;
  durationBucketMs?: number | null;
  failureCountBucket?: string | null;
  createdAt?: string | null;
}

export interface ProviderQuotaWindowItem {
  provider: string;
  providerCategory?: string | null;
  routeFamily?: string | null;
  windowType: string;
  windowStart: string;
  windowEnd: string;
  requestCount: number;
  reservedUnits: number;
  consumedUnits: number;
  releasedUnits: number;
  rejectedCount: number;
  successCount: number;
  failureCount: number;
  timeoutCount: number;
  provider429Count: number;
  provider403Count: number;
  fallbackCount: number;
  probeCount: number;
  cacheOnlyCount: number;
  staleServedCount: number;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface ProviderProbeEventItem {
  provider: string;
  providerCategory?: string | null;
  routeFamily?: string | null;
  probeType: string;
  probeSource: string;
  resultBucket: string;
  durationBucketMs?: number | null;
  createdAt?: string | null;
}

export interface ProviderRecentErrorBucketItem {
  reasonBucket: string;
  countBucket: string;
  latestAt?: string | null;
}

export interface ProviderSlaTrendSummaryItem {
  windowCountBucket: string;
  requestCountBucket: string;
  failureCountBucket: string;
  timeoutCountBucket: string;
  provider429CountBucket: string;
  provider403CountBucket: string;
  latestObservationAt?: string | null;
}

export interface ProviderSlaReadinessItem {
  provider: string;
  providerCategory?: string | null;
  routeFamily?: string | null;
  readinessState: string;
  reasonCode: string;
  credentialState: string;
  liveProvidersEnabled: boolean;
  providerEnabled: boolean;
  credentialsPresent: boolean;
  dryRunEnabled: boolean;
  liveHttpCallsEnabled: boolean;
  brokerOrderPathEnabled: boolean;
  portfolioMutationPathEnabled: boolean;
  tradeableData: boolean;
  latencyBucketMs?: number | null;
  latencyState: string;
  errorRate?: number | null;
  errorState: string;
  freshnessSeconds?: number | null;
  freshnessState: string;
  recentErrors: ProviderRecentErrorBucketItem[];
  trendSummary?: ProviderSlaTrendSummaryItem;
  circuitAdvisoryState: string;
  circuitStateCandidate: string;
  liveEnforcement: boolean;
  wouldBlockCall: boolean;
  wouldChangeProviderOrder: boolean;
  wouldChangeFallbackBehavior: boolean;
  noExternalCalls: boolean;
}

export interface ProviderCircuitResponse<T> {
  generatedAt: string;
  items: T[];
  metadata: ProviderCircuitDiagnosticsMetadata;
}

export interface ProviderCircuitDiagnosticsBundle {
  states: ProviderCircuitResponse<ProviderCircuitStateItem>;
  events: ProviderCircuitResponse<ProviderCircuitEventItem>;
  quotaWindows: ProviderCircuitResponse<ProviderQuotaWindowItem>;
  probeEvents: ProviderCircuitResponse<ProviderProbeEventItem>;
  slaReadiness: ProviderCircuitResponse<ProviderSlaReadinessItem>;
}

function safeArray<T>(value: unknown): T[] {
  return Array.isArray(value) ? value as T[] : [];
}

function safeLimit(limit?: number): number {
  const parsed = Number(limit || 50);
  if (!Number.isFinite(parsed)) return 50;
  return Math.min(Math.max(Math.trunc(parsed), 1), 200);
}

function safeLabel(value?: string): string | undefined {
  const normalized = String(value || '').trim().toLowerCase().replace(/[^a-z0-9:_-]/g, '').slice(0, 64);
  return normalized || undefined;
}

function toSafeQuery(params: ProviderCircuitDiagnosticsParams = {}, extras: Record<string, string | undefined> = {}) {
  return {
    provider: safeLabel(params.provider),
    routeFamily: safeLabel(params.routeFamily),
    since: typeof params.since === 'string' ? params.since.slice(0, 64) : undefined,
    limit: safeLimit(params.limit),
    ...extras,
  };
}

function normalizeMetadata(value?: Partial<ProviderCircuitDiagnosticsMetadata> | null): ProviderCircuitDiagnosticsMetadata {
  return {
    readOnly: value?.readOnly === true,
    noExternalCalls: value?.noExternalCalls === true,
    liveEnforcement: value?.liveEnforcement === true,
    providerBehaviorChanged: value?.providerBehaviorChanged === true,
    marketCacheBehaviorChanged: value?.marketCacheBehaviorChanged === true,
    dataSources: safeArray<string>(value?.dataSources),
    limit: Number(value?.limit || 50),
    redaction: safeArray<string>(value?.redaction),
    filters: value?.filters || {},
  };
}

function normalizeResponse<T>(payload: Record<string, unknown>): ProviderCircuitResponse<T> {
  const normalized = toCamelCase<ProviderCircuitResponse<T>>(payload);
  return {
    generatedAt: normalized.generatedAt || '',
    items: safeArray<T>(normalized.items),
    metadata: normalizeMetadata(normalized.metadata),
  };
}

export const adminProviderCircuitsApi = {
  async getStates(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitResponse<ProviderCircuitStateItem>> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/providers/circuits', {
      params: toSafeQuery(params, { state: safeLabel(params.state) }),
    });
    return normalizeResponse<ProviderCircuitStateItem>(response.data);
  },

  async getEvents(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitResponse<ProviderCircuitEventItem>> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/providers/circuits/events', {
      params: toSafeQuery(params, { eventType: safeLabel(params.eventType) }),
    });
    return normalizeResponse<ProviderCircuitEventItem>(response.data);
  },

  async getQuotaWindows(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitResponse<ProviderQuotaWindowItem>> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/providers/quota-windows', {
      params: toSafeQuery(params),
    });
    return normalizeResponse<ProviderQuotaWindowItem>(response.data);
  },

  async getProbeEvents(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitResponse<ProviderProbeEventItem>> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/providers/probe-events', {
      params: toSafeQuery(params),
    });
    return normalizeResponse<ProviderProbeEventItem>(response.data);
  },

  async getSlaReadiness(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitResponse<ProviderSlaReadinessItem>> {
    const response = await apiClient.get<Record<string, unknown>>('/api/v1/admin/providers/sla-readiness', {
      params: toSafeQuery(params),
    });
    return normalizeResponse<ProviderSlaReadinessItem>(response.data);
  },

  async getDiagnostics(params: ProviderCircuitDiagnosticsParams = {}): Promise<ProviderCircuitDiagnosticsBundle> {
    const [states, events, quotaWindows, probeEvents, slaReadiness] = await Promise.all([
      adminProviderCircuitsApi.getStates(params),
      adminProviderCircuitsApi.getEvents(params),
      adminProviderCircuitsApi.getQuotaWindows(params),
      adminProviderCircuitsApi.getProbeEvents(params),
      adminProviderCircuitsApi.getSlaReadiness(params),
    ]);
    return { states, events, quotaWindows, probeEvents, slaReadiness };
  },
};
